demo/kitchen-sink/demo.js

```diff
@@ -103,29 +81,104 @@ require("ace/config").defineOptions(Editor.prototype, "editor", {
+                    className: el.kind == 3
+                        ? "language_highlight_occurrence_main"
+                        : "language_highlight_occurrence_other"
+                };
+            }));
+        }
+        function showAnnotations(session, diagnostics) {
+            session.clearAnnotations();
+            let annotations = diagnostics.map((el) => {
+                console.log(el.severity, el)
+                return {
+                    row: el.range.start.line,
+                    column: el.range.start.character,
+                    text: el.message,
+                    type: el.severity === 1 ? "error" : el.severity === 2 ? "warning" : "info"
+                };
+            });
+            if (annotations && annotations.length > 0) {
+                session.setAnnotations(annotations);
+            }
+            
+            if (!session.state) session.state = {}
+            if (!session.state.diagnosticMarkers) {
+                session.state.diagnosticMarkers = new MarkerGroup(session);
+            }
+            session.state.diagnosticMarkers.setMarkers(diagnostics.map(function(el) {
+                var r = el.range;
+                return {
+                    range: new Range(r.start.line, r.start.character, r.end.line, r.end.character),
+                    tooltipText: el.message,
+                    className:  "language_highlight_error"
+                };
+            }));
+        };
         
         docTooltip.setDataProvider(function(e, editor) {
-            var renderer = editor.renderer;
-
             let session = editor.session;
-            let docPos = e.getDocumentPosition() ;
+            let docPos = e.getDocumentPosition();
 
             languageProvider.doHover(session, docPos, function(hover) {
-                if (!hover) {
-                    return;
-                }
-                // todo should ace itself handle markdown?
-                var domNode = dom.buildDom(["p", {}, hover.content.text]);
-                docTooltip.showForRange(editor, hover.range, domNode, e);
+                var errorMarker = session.state?.diagnosticMarkers.getMarkerAtPosition(docPos);
+                var range = hover?.range || errorMarker?.range;
+                if (!range) return;
+                var hoverNode = hover && dom.buildDom(["div", {}])
+                if (hoverNode) {
+                    hover.content.text = hover.content.text.replace(/(?!^)`{3}/gm, "\n$&");
+                    // todo render markdown using ace markdown mode
+                    hoverNode.innerHTML = languageProvider.getTooltipText(hover);
+                };
+                
+                var domNode = dom.buildDom(["div", {},
+                    hoverNode,
+                    errorMarker && ["div", {}, errorMarker.tooltipText.trim()]
+                ]);
+                docTooltip.showForRange(editor, range, domNode, e);
```

Comment:
looking at this example, we provide only the separate building blocks (marker group to display markers and hover tooltip to display tooltips), but no facilities to connect them - meaning the integrator of Ace would need to combine those however they see fit. it might make sense, since it provides more flexibility and control over the behaviour (like here we combine hover and diagnostics in one tooltip), but in this case is there any sense in including `tooltipText` in the marker type definition? it's purpose there is far from obvious, and integrators should be able to extract that information outside Ace with equal success. curious of your opinion @andredcoliveira

Reply:
it would also be great to add a note to our documentation (e.g. examples on wiki) how to use the new additions (hover tooltip, marker group)